test/api/MouseTracking.api.ts

```diff
@@ -0,0 +1,2021 @@
+    (function() {
+      const rect = window.term.element.getBoundingClientRect();
+      const dim = term._core._renderService.dimensions;
+      return {left: rect.left, top: rect.top, bottom: rect.bottom, right: rect.right, width: dim.actualCellWidth, height: dim.actualCellHeight};
+    })();
+  `);
+  return [col * coords.width + coords.left + 2, row * coords.height + coords.top + 2];
+}
+
+/**
+ * Patched puppeteer functions.
+ * This is needed to:
+ *  - translate cell positions into pixel positions
+ *  - allow modifiers to be set
+ *  - fake wheel events
+ */
+async function mouseMove(col: number, row: number): Promise<void> {
+  const [xPixels, yPixels] = await cellPos(col, row);
+  return await page.mouse.move(xPixels, yPixels);
+}
+async function mouseClick(col: number, row: number): Promise<void> {
+  const [xPixels, yPixels] = await cellPos(col, row);
+  return await page.mouse.click(xPixels, yPixels);
+}
+async function mouseDown(button: "left" | "right" | "middle" | undefined): Promise<void> {
+  return await page.mouse.down({button});
+}
+async function mouseUp(button: "left" | "right" | "middle" | undefined): Promise<void> {
+  return await page.mouse.up({button});
+}
+async function wheelUp(): Promise<void> {
+  const self = (page.mouse as any);
+  return await page.evaluate(`
+    window.term.element.dispatchEvent(new WheelEvent('wheel', {clientX: ${self._x}, clientY: ${self._y}, deltaX: 0, deltaY: -10, modifiers: ${self._keyboard._modifiers}}));
+  `);
+}
+async function wheelDown(): Promise<void> {
+  const self = (page.mouse as any);
+  return await page.evaluate(`
+    window.term.element.dispatchEvent(new WheelEvent('wheel', {clientX: ${self._x}, clientY: ${self._y}, deltaX: 0, deltaY: 10, modifiers: ${self._keyboard._modifiers}}));
+  `);
+}
+
+// button definitions
+const buttons: {[key: string]: number} = {
+  '<none>':  -1,
+  left:       0,
+  middle:     1,
+  right:      2,
+  released:   3,
+  wheelUp:    4,
+  wheelDown:  5,
+  wheelLeft:  6,
+  wheelRight: 7,
+  aux8:       8,
+  aux9:       9,
+  aux10:      10,
+  aux11:      11,
+  aux12:      12,
+  aux13:      13,
+  aux14:      14,
+  aux15:      15,
+}
+const reverseButtons: any = {};
+for (let el in buttons) {
+  reverseButtons[buttons[el]] = el;
+}
+
+// extract button data from buttonCode
+function evalButtonCode(code: number) : any {
+  const modifier = {shift: !!(code & 4), meta: !!(code & 8), control: !!(code & 16)};
+  const move = code & 32;
+  let button = 0;
+  if (code  === 3) {
+    button = 3;
+    code -= 3;
+  } else {
+    if (code >= 128) {
+      button |= 8;
+      code -= 128;
+    }
+    if (code >= 64) {
+      button |= 4;
+      code -= 64;
+    }
+    button += code & 3;
+  }
+  let actionS = 'press';
+  let buttonS = reverseButtons[button];
+  if (move) {
+    actionS = 'move';
+  } else if (button === 3) {
+    buttonS = '<none>';
+    actionS = 'release';
+  } else if ((button & 4) && !(button & 8)) {
```

Comment:
I think directly using the button numbers would be clearer:
```suggestion
  } else if (4 <= button && button <= 7) {
```

For `actionS` the bit-ops could be replaced by simple `button == 4` etc or even simply indexing into `buttons`

Reply:
Yepp, simplified it abit.